### DISCUSSION TITLE
Adopt `Suspense` for `useLiveQuery`

### DISCUSSION THREAD

**Discussion by FuJuntao** [dexie/Dexie.js#1276] [Ideas]
Hi, I see there is no property like `isLoading` for `useLiveQuery` result.

I wonder if it is possible to adopt `Suspense` for `useLiveQuery`.

**Comment by dfahlander** [comment]
isLoading: The third argument to [useLiveQuery()](https://dexie.org/docs/dexie-react-hooks/useLiveQuery()) is a default value returned while loading. With IndexedDB, loading time is normally too short to make it feasible with fallback rendering. I normally just return null from the component while loading.

Examples:

```jsx
export function FriendListComponent() {
  // When returning array, we can rely on the default and do not need to specify 3rd arg:
  const friends = useLiveQuery(() => db.friends.toArray());

  // A falsy result means we're still loading
  if (!friends) return null; // Initial render (is loading)

  return <ul>{
    friends.map(friend => <li key={friend.id}>{friend.name}</li>)
  }</ul>;
}

export function FriendDetailsComponent(friendId) {
   const friend = useLiveQuery(
     () => db.friends.get(friendId), // This query may return undefined!
     [friendId], // Deps
     null // Specify fallback distinguished from not-found result.
   );

   if (friend === null) return null; // Initial render (is loading)

   if (friend === undefined) return <p>Friend not found</p>; // Not found.

   return <div>
     <p>Name: {friend.name}</p>
     <p>Age: {friend.age}</p>
   </div>;
}
```

Suspense support: We might support a suspense based useLiveQuery in future. That would require a DBCore caching middleware that is capable of caching all responses including IDBCoreCursors - which in turn would require a virtual cursor ("CachedCursor" or how to name it) to be implemented that will behave just like the original cursor. So, it's doable but not super straightforward to implement.

**Reply by FuJuntao** [reply]
Sounds like a lot of thoughts and works, but I still think it's a nice to have feature. Hopefully there will be time for implementing this.

**Comment by bioshazard** [comment]
The default return worked well for me. I used a `Symbol` tho in case there are an cases where Dexie would legitimately return `null`. eg,

```
import { useLiveQuery } from 'dexie-react-hooks';
import db from 'lib/db';

// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol
const dexieLoading = Symbol()

export default function useSettingDexie({ name, defaultValue }) {
  const liveQuery = useLiveQuery(
    () => db.settings.get({ name }),
    [name],
    dexieLoading // default to unique "loading" symbol
  )
  const loading = liveQuery === dexieLoading
  const value = loading ? undefined // return `undefined` if still loading...
    : liveQuery === undefined // after loading, if query is undefined... it means 404 from Dexie
      ? defaultValue // relay the default value
      : liveQuery.value // otherwise, relay the query value itself

  const save = async (value) => await db.settings.put({ name, value });

  return {
    liveQuery,
    loading,
    value,
    save
  }
}
```

**Comment by mutewinter** [comment]
Here's my approach for adding Suspense support to `liveQuery` using [Jotai](https://github.com/pmndrs/jotai). I also added a cache to ensure the minimal amount of observables are created.

```typescript
// utils/useSuspenseLiveQuery.ts
import { liveQuery } from 'dexie';
import { type Atom, useAtomValue } from 'jotai';
import { atomWithObservable } from 'jotai/utils';
import { useRef } from 'react';

// Must use a global cache because Suspense breaks the rules of hooks like
// useMemo and useRef
// https://github.com/facebook/react/issues/15137
// eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-redundant-type-constituents
const CACHE = new Map<string, Atom<any | Promise<any>> | undefined>();

// Cache for 5 minutes, which was inspired by TanStack Query
const CACHE_ITEM_LIFESPAN_MS = 1000 * 60 * 5;

type CacheKey = string | number;
// Force at least one cache key past the `use` prefixed key to ensure that
// using useSuspenseLiveQuery is necessary vs a global atom.
type CacheKeys<T = CacheKey> = [`use${string}`, T, ...T[]];

type InitialValue<Data> = Data | (() => Data);

// A suspense-enabled version of useLiveQuery from Dexie
export default function useSuspenseLiveQuery<T>(
  querier: () => Promise<T> | T,
  cacheKeys: CacheKeys,
): T | undefined;
export default function useSuspenseLiveQuery<T, U extends InitialValue<T>>(
  querier: () => Promise<T> | T,
  cacheKeys: CacheKeys,
  initialValue: U,
): T | U;
export default function useSuspenseLiveQuery<T, U extends InitialValue<T>>(
  querier: () => Promise<T> | T,
  cacheKeys: CacheKeys,
  initialValue?: U,
): T | U {
  const cacheKey = cacheKeys.join(':');
  const atomRef = useRef<[string, Atom<T | Promise<T>> | undefined]>([
    cacheKey,
    CACHE.get(cacheKey),
  ]);

  // Look for atom in cache
  if (atomRef.current[0] !== cacheKey) {
    atomRef.current = [cacheKey, CACHE.get(cacheKey)];
  }

  // Atom not in cache
  if (!atomRef.current[1]) {
    const newAtom = atomWithObservable<T>(
      () => liveQuery(querier),
      initialValue ? { initialValue } : undefined,
    );
    if (process.env.NODE_ENV === 'development') {
      newAtom.debugLabel = cacheKey;
    }
    atomRef.current = [cacheKey, newAtom];
    CACHE.set(cacheKey, newAtom);

    // Remove from cache after a timeout to allow for garbage collection
    setTimeout(() => {
      CACHE.delete(cacheKey);
    }, CACHE_ITEM_LIFESPAN_MS);
  }

  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  return useAtomValue(atomRef.current[1]!);
}
```

## Example Use

```typescript
// store/sessions.ts
export function useSession(sessionId: SessionId) {
  return useSuspenseLiveQuery(
    () => db.sessions.get(sessionId),
    ['useSession', sessionId],
  );
}
```

**Reply by mutewinter** [reply]
You can also use Jotai for atoms that don't need any params like so:

```typescript
import { liveQuery } from 'dexie';
import { useAtomValue } from 'jotai';
import { atomFamily, atomWithObservable } from 'jotai/utils';

const sessionsByUpdatedAtAtom = atomWithObservable(() =>
  liveQuery(() => db.sessions.orderBy('updatedAt').reverse().toArray()),
);

export function useSessionsByUpdatedAt() {
  return useAtomValue(sessionsByUpdatedAtAtom);
}
```

**Reply by dfahlander** [reply]
I must have missed this suggestion. It's a nice take on it.

**Comment by phaux** [comment]
@dfahlander Now that [view transitions](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_view_transitions) are stable this is even more important. Without suspense there is a transition to the empty state and then the data appears only after the view transition finishes. This only makes things worse because normally the empty state shows only for split second. Ideally there would be no empty state visible at all.

**Reply by dfahlander** [reply]
Thanks for the explanation. @mutewinter's suggestion would probably be an easy way to get this right now. The only thing stopping me from including such hook in dexie-react-hooks is the dream of having a suspense version without a cache-key - which would be possible but a bit harder to solve and might require a slightly different approach - with a sync dexie api. It's probably worth it to introduce a useSuspenseLiveQuery() in the meantime. Please try @mutewinter's hook out and give feedback.

**Reply by phaux** [reply]
Seems to be working. I wonder if it can be achieved without using jotai because I don't need it.

**Reply by dfahlander** [reply]
> Seems to be working. I wonder if it can be achieved without using jotai because I don't need it.

Yes, I think this could be fixed without Jotai with not so much code. We could throw a Promise when item is not in cache - a Promise that resolves when the observable emits initially.

Anyone eager to create a PR in https://github.com/dexie/Dexie.js/tree/master/libs/dexie-react-hooks are much welcome. Otherwise I will jump into it when there is time... (which would probably be in September or so).

I think we could build a `useSuspenseObservable()` first that does the same as useObservable() but with suspense and build useSuspenseLiveQuery() on top of it just like useLiveQuery() is built on top of useObservable().

Thanks @mutewinter for the example code. I'd be certain to mention you.

**Reply by mutewinter** [reply]
Thank you for the incredible library. Here’s the codebase where I heavily use that Jotai-based observable hook: https://github.com/mutewinter/dreams/blob/main/src/utils/useSuspenseLiveQuery.ts

**Comment by phaux** [comment]
I made a WIP PR: #2205

**Comment by dpnova** [comment]
Not sure if it's related to this, but I haven't been able to find a clean way to use dexie promise api with the new use function in react. Using a memo on the promise doesn't seem to be stable enough.

Edit: the reason why I ask is it seems having "suspense enabled data fetch" is handy with the new Activity component in react 19.2.

We already use useLiveQuery heavily...

**Comment by dfahlander** [comment]
Hi everyone, please give [dexie-react-hooks@4.2.1-beta.1](https://www.npmjs.com/package/dexie-react-hooks/v/4.2.1-beta.1) a try, with [useSuspendingLiveQuery](https://github.com/dexie/Dexie.js/blob/master/libs/dexie-react-hooks/src/useSuspendingLiveQuery.ts) by [Niki](https://github.com/phaux) if curios on a suspense version of live queries.

All feedback is welcome!

One difference from the ordinary useLiveQuery() is that it not only suspends on initial load but also when deps change, so should probably be used together with React's useTransition(), (whilst useLiveQuery() keeps rendering the last state until the new results are available).
